sphinx_immaterial/apidoc/cpp/api_parser.py

```diff
@@ -429,68 +441,60 @@ def _get_all_decls(config: Config, cursor: Cursor, allow_file):
             yield from _get_all_decls(config, child, None)
 
 
+NON_DOC_COMMENT = re.compile(
+    r"(^//[^/\!].*$)|(^/\*[^\*\!](?:.|\n)*?\*/$)", re.MULTILINE
```

Comment:
I think if we have `/*` inside of a `///` doc comment, it will also get stripped out.  Also, if we just delete lines, then the line numbers won't be correct.  We could either change to specifying separate location info for each line, or could replace each removed line with an empty line.

Additionally, the logic below for stripping out the comment syntax assumes just a single comment type.

I think to properly handle multiple comment types a different strategy may be required here:
1. Parse `cmt` into a sequence of comments and whitespace (can be done with a regular expression).
2. Coalesce consecutive doc comment lines that start with `///` or `//!`.  I think you could actually design the regular expression so that this can be done as part of step 1.
3. Remove any non-doc comments, including the comment begin/end marker, but preserve any line breaks to avoid messing up the line numbers.
4. For each coalesced doc comment, strip the comment begin/end marker and dedent the text.

Reply:
For `//` comments that get stripped, the line break should still remain, although `dedent()` may be stripping empty lines (I can't remember what the docs said about that).

For `/* ... */` comments that get stripped, only the final line break is preserved. It would take multiple regex patterns to preserve the line breaks for comments that span multiple lines (see the last test in `test_comment_styles()`).

I'd like to say it is easier to add empty lines to preserve the src's line number, but maybe not. Adjusting the lines' number for the src seems harder, but I haven't looked into that yet.

The steps you propose might require an approach that doesn't use `decent()`, which I opted for out of convenience.

In my research, I translated the stripped `raw_comment` line-by-line while keeping track of the indentation depth -- it wasn't perfect for `@code\n ... \n ... \n@endcode` blocks, but it worked. I was hoping to refactor the comment stripping/parsing out of api_parse.py before taking a line-by-line approach.

Reply:
I think it will be possible to use `dedent`, we just need to apply it after splitting by comment type.

Once you split by comment type it is also easy to remove a comment while preserving newlines.  Just remove all characters except `\n`.

Reply:
I split the raw_comment into chunks based on the `re.search()` result's `start/end()`. Now it strips the sections that are docstrings (1 at a time), replaces the non-doc comments with blank lines (determined by `match.group().count("\n")`), and finally dedents the striped docstrings.

I think this should satisfy the concern of keeping the src's line numbers consistent in any sphinx warnings.